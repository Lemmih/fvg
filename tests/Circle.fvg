data List a = Nil | Cons a (List a)
data Tuple a b = Tuple a b

enumerate : Int -> Int -> (Int -> a) -> List a
enumerate n max fn =
  case ltI n max of
    False -> Nil
    True -> Cons (fn n) (enumerate (plusI n 1) max fn)

mkCoords : Int -> Float -> List (Tuple Float Float)
mkCoords n radius =
  let width = plus (times radius 2.0) 50.0 in
  enumerate 0 n (\i ->
    let angle = times (div (itof i) (div (itof n) 2.0)) pi in
    let x = plus (times radius (cos angle)) (div width 2.0) in
    let y = plus (times radius (sin angle)) (div width 2.0) in
    Tuple x y
   )

for : List a -> (a -> b) -> List b
for lst fn =
  case lst of
    Nil -> Nil
    Cons x xs -> Cons (fn x) (for xs fn)

main : Svg
main =
  let radius = 200.0 in
  let size = plus (times radius 2.0) 50.0 in
  <svg width="{{size}}" height="{{size}}">
    for (mkCoords 10 radius) (\tuple ->
      case tuple of
        Tuple x y ->
          <circle r="5" cx="{{x}}" cy="{{y}}"> </circle>
      )
  </svg>
