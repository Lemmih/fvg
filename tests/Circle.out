 <svg width="450.0" height="450.0" > <circle r="5" cx="425.0" cy="225.0" > </circle>  <circle r="5" cx="386.8033988749895" cy="342.55705045849464" > </circle>  <circle r="5" cx="286.8033988749895" cy="415.21130325903073" > </circle>  <circle r="5" cx="163.19660112501055" cy="415.21130325903073" > </circle>  <circle r="5" cx="63.196601125010545" cy="342.55705045849464" > </circle>  <circle r="5" cx="25.0" cy="225.00000000000003" > </circle>  <circle r="5" cx="63.19660112501049" cy="107.4429495415054" > </circle>  <circle r="5" cx="163.1966011250105" cy="34.7886967409693" > </circle>  <circle r="5" cx="286.80339887498945" cy="34.78869674096927" > </circle>  <circle r="5" cx="386.80339887498945" cy="107.44294954150533" > </circle>  </svg> 